### PR TITLE
Adding StringUtils class

### PR DESCRIPTION
**Pull request by streetsofboston**
Adding a bunch of static string manipulation methods in the StringUtils class.

### CODE REVIEW

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+    private static final ThreadLocal<Pattern[]> MATCHING_PATTERNS = new ThreadLocal<Pattern[]>() {
+        @Override
+        protected Pattern[] initialValue() {
+            return new Pattern[] {
+                    Pattern.compile("[\\t\\n\\u0590-\\uFFFF]"), // non-boring characters.
+                    Pattern.compile("[^a-zA-Z0-9]"), // non-alpha-numeric.
+                    Pattern.compile("^([a-z0-9A-Z]+[-_|\\.]?)+[a-z0-9A-Z]@([a-z0-9A-Z]+(-[a-z0-9A-Z]+)?\\.)+[a-zA-Z]{2,}$") // Email-address pattern.
+            };
+        }
+    };
```
Comment (russhwolf) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java:38]: Is there a reason this is an array instead of having three separate constants?

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+     * @param string      String whose non-boring character will be replace with the given replacement.
+     * @param replacement The replacement.
+     * @return The string without boring characters.
+     * @see android.text.BoringLayout
+     * <p/>
+     * Boring characters are all character less than \u0590 and not a tab
+     * and not a new line.
+     */
+    public static String replaceNonBoringCharacters(@NonNull CharSequence string, @NonNull String replacement) {
+        return MATCHING_PATTERNS.get()[0].matcher(string).replaceAll(replacement);
```
Comment (russhwolf) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java]: Might be worth adding error checking in here to enforce that the replacement string only contains non-boring characters.

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java (hunk 1 above)
Comment (streetsofboston) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java:38]: ThreadLocals are relatively expensive. Better have one instead of more than one. I was debating to have it return an object with three Patterns or an 3-Pattern array. I went with the last one.

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java (hunk 2 above)
Comment (streetsofboston) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java]: This method would be mainly used to strip out non-boring character (i.e. replacement==""). 

The caller should make sure that the replacement is non-boring as well. It can be a little expensive to check for this and we should avoid it if the caller already knows that the replacement doesn't contain any non-boring chars.

I will add some null/empty string checks, though.

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+     * The returned string will never start or end with the given separator unless the input string is empty.
+     *
+     * @param string      The string to repeat.
+     * @param separator   The separator between repetitions.
+     * @param repeatCount The times that the given string should be repeated.
+     * @return The repeated string.
+     */
+    public static CharSequence repeat(CharSequence string, CharSequence separator, int repeatCount) {
+        if (repeatCount < 1) {
+            return null;
```
Comment (russhwolf) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java]: Not sure if I have a strong opinion, but would it be better to pass an empty string instead of null here?

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+                // matching current letters in both strings
+                int match = (string1.charAt(i - 1) == string2.charAt(j - 1)) ? 0 : 1;
+
+                // computing cost for each transformation
+                int cost_replace = cost[i - 1] + match;
+                int cost_insert = cost[i] + 1;
+                int cost_delete = newcost[i - 1] + 1;
+
+                // keep minimum cost
+                newcost[i] = Math.min(Math.min(cost_insert, cost_delete), cost_replace);
```
Comment (russhwolf) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java:341]: I wonder if we need a MathUtils.min()/max() with a variable number of inputs :)

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+    public static boolean equals(CharSequence string1, CharSequence string2) {
+        return compare(string1, string2) == 0;
+    }
+
+    /**
+     * Has the same function as {@link java.lang.String#compareTo(String)} but it works on CharSequences as well.
+     *
+     * @param cs1 First string.
+     * @param cs2 Second string.
+     * @return -1, 0, or 1.
```
Comment (russhwolf) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java]: As written, this returns negative, zero, or positive depending on the comparison, but not always -1, 0, or 1.

Comment (streetsofboston) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java]: Woops. JavaDoc needs to be fixed here. Good catch.

File: commonutils/src/test/java/io/intrepid/commonutils/StringUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    public void testStringToEnumFail() {
+        try {
+            assertEquals(TestEnum.TEST_ENUM2, StringUtils.toEnum(TestEnum.class, "HELLO"));
+        }
+        catch (IllegalArgumentException iae) {
+            // OK
+            return;
+        }
+
+        fail("Excpected an IllegalArgumentException");
```
Comment (russhwolf) [inline commonutils/src/test/java/io/intrepid/commonutils/StringUtilsTest.java]: Typo Excpected -> Expected

**Comment by russhwolf** [comment]
Went through and left a few comments. Seems good for the most part, although there's a number of things in here that I've never personally had need of so I don't know if I'm fully considering all uses as I read through them.

File: commonutils/src/test/java/io/intrepid/commonutils/util/MockedTextUtils.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * Mocks the {@link TextUtils} class.
+ *
+ * Created by aspaans on 5/6/16.
+ */
+public class MockedTextUtils {
+    public static void setup() {
+        PowerMockito.mockStatic(TextUtils.class);
+        Mockito.when(TextUtils.isEmpty(any(CharSequence.class))).then(new Answer<Boolean>() {
+            @Override
```
Comment (f2rf2r) [inline commonutils/src/test/java/io/intrepid/commonutils/util/MockedTextUtils.java:21]: let's just make isEmpty a method in our StringUtils class. We should move away from using TextUtils.isEmpty() since it's a pain to get it to work in unit tests

File: commonutils/build.gradle [hunk 8]
```diff
@@ -29,8 +29,19 @@ apply from: 'publish.gradle'
...
     testCompile 'junit:junit:4.12'
+    // Set this dependency if you want to use Mockito
     testCompile 'org.mockito:mockito-all:1.10.19'
+    // Set this dependency if you want to use Hamcrest matching
+    testCompile 'org.hamcrest:hamcrest-library:1.1'
+    // Powermock
+    testCompile "org.powermock:powermock-module-junit4:1.6.4"
+    testCompile "org.powermock:powermock-module-junit4-rule:1.6.4"
+    testCompile "org.powermock:powermock-api-mockito:1.6.4"
+    testCompile "org.powermock:powermock-classloading-xstream:1.6.4"
```
Comment (f2rf2r) [inline commonutils/build.gradle]: extract the version into a varaible

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+
+/**
+ * A utility class bundling a set of handy to use methods that deal with {@link String},
+ * {@link CharSequence}, {@link android.text.Spannable}, etc.
+ * <p/>
+ * Created by Anton Spaans on 4/15/16.
```
Comment (f2rf2r) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java]: remove the author info

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+ */
+@SuppressWarnings("unused")
+public abstract class StringUtils {
```
Comment (f2rf2r) [inline commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java:28]: rename to IntrepidStringUtils

Comment (streetsofboston) [reply commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java:28]: Why? That is just too wordy. The package name already makes sure the name is scoped properly.

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java (hunk 4 above)
Comment (streetsofboston) [reply commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java:341]: Yes. We should replace this line with a `clamp` or some similar method when a MathUtils is added.

File: commonutils/src/test/java/io/intrepid/commonutils/util/MockedTextUtils.java (hunk 7 above)
Comment (streetsofboston) [reply commonutils/src/test/java/io/intrepid/commonutils/util/MockedTextUtils.java:21]: Simple mock helps with this issue.

**Comment by streetsofboston** [comment]
I addressed the PR comments. Let me know what you think.

File: commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java (hunk 10 above)
Comment (f2rf2r) [reply commonutils/src/main/java/io/intrepid/commonutils/StringUtils.java:28]: The apps using this library will likely have their own StringUtils class since this class won't cover all use cases, and there's a good chance both of these classes will end up being used in the same file. Even though the names are scoped by package, they will end up writing out the full path for one of these class, which leads to ugly code. It also leads to confusion when ide shows the autocompletion suggestions.

File: commonutils/src/test/java/io/intrepid/commonutils/util/MockedTextUtils.java (hunk 7 above)
Comment (f2rf2r) [reply commonutils/src/test/java/io/intrepid/commonutils/util/MockedTextUtils.java:21]: It's more for the apps using this library than testing our lib. It's simple to mock it here, but we would need to redo the mock for every app that uses and tests around TextUtils.isEmpty() . Using powermock is annoying and usually is a sign of bad code design. For such a trivial method, we should just redefine it rather than relying on the TextUtils.

File: commonutils/build.gradle [hunk 11]
```diff
@@ -27,10 +28,31 @@ ext.mavenGroupId = 'io.intrepid.commonutils'
...
-    testCompile 'org.mockito:mockito-all:1.10.19'
-    testCompile "org.robolectric:robolectric:3.0"
-    compile 'com.android.support:appcompat-v7:24.1.1'
-}
\ No newline at end of file
+    compile "com.android.support:appcompat-v7:$supportVersion"
+
+    //Rx
+    compile "io.reactivex:rxjava:$rxJavaVersion"
+    compile "io.reactivex:rxandroid:$rxAndroidVersion"
```
Comment (f2rf2r) [inline commonutils/build.gradle:45]: why are we including Rx and retrolambda in this library?

**Review by f2rf2r — APPROVED** [review]
